File: packages/protocol/src/receiver.ts
/**
 * Receiver rules for signaling envelopes (spec section 6), applied in order.
 * Every failure is final for that message. The Swift and Kotlin implementations
 * mirror this class and share its test vectors.
 */
import { b64urlDecode, utf8Decode, utf8Encode } from './encoding.ts';
import {
  MAX_CLOCK_SKEW_MS,
  MAX_ENVELOPE_BYTES,
  SUPPORTED_VERSIONS,
  decodePayload,
  isSignalingType,
  verifyEnvelopeSignature,
  type Envelope,
} from './envelope.ts';
import { deviceIdFromPublicKey, importPublicKey } from './identity.ts';
import { validateEnvelopeShape, validateSignalingPayload } from './validate.ts';

export type RejectReason =
  | 'too_large'
  | 'malformed'
  | 'unsupported_version'
  | 'wrong_recipient'
  | 'unknown_type'
  | 'unknown_sender'
  | 'bad_signature'
  | 'stale_timestamp'
  | 'replayed'
  | 'invalid_payload';

export interface Accepted {
  ok: true;
  envelope: Envelope;
  payload: unknown;
  /** The key that verified the signature. For PAIR_REQUEST this is the key carried in the payload. */
  senderKey: CryptoKey;
}

export interface Rejected {
  ok: false;
  reason: RejectReason;
  detail?: string;
}

export type ReceiveResult = Accepted | Rejected;

export type KeyResolver = (deviceId: string) => Promise<CryptoKey | null> | CryptoKey | null;

export interface ReceiverOptions {
  selfDeviceId: string;
  /** Returns the public key of a trusted device, or null if unknown or revoked. */
  resolveKey: KeyResolver;
  now?: () => number;
  maxSkewMs?: number;
  /** Only the host while a pairing session is open. Controllers never accept PAIR_REQUEST. */
  acceptPairRequests?: boolean;
  supportedVersions?: readonly number[];
}

function reject(reason: RejectReason, detail?: string): Rejected {
  return detail === undefined ? { ok: false, reason } : { ok: false, reason, detail };
}

function seqKey(from: string, session: string): string {
  return `${from}|${session}`;
}

export class EnvelopeReceiver {
  readonly #self: string;
  readonly #resolveKey: KeyResolver;
  readonly #now: () => number;
  readonly #maxSkew: number;
  readonly #acceptPair: boolean;
  readonly #versions: readonly number[];
  readonly #lastSeq = new Map<string, number>();

  constructor(opts: ReceiverOptions) {
    this.#self = opts.selfDeviceId;
    this.#resolveKey = opts.resolveKey;
    this.#now = opts.now ?? (() => Date.now());
    this.#maxSkew = opts.maxSkewMs ?? MAX_CLOCK_SKEW_MS;
    this.#acceptPair = opts.acceptPairRequests ?? false;
    this.#versions = opts.supportedVersions ?? SUPPORTED_VERSIONS;
  }

  async receive(raw: string | Uint8Array): Promise<ReceiveResult> {
    const bytes = typeof raw === 'string' ? utf8Encode(raw) : raw;
    if (bytes.length > MAX_ENVELOPE_BYTES) return reject('too_large');

    let parsed: unknown;
    try {
      parsed = JSON.parse(utf8Decode(bytes));
    } catch {
      return reject('malformed', 'not json');
    }
    const shape = validateEnvelopeShape(parsed);
    if (!shape.valid) return reject('malformed', shape.errors[0]);
    const env = parsed as Envelope;

    if (!this.#versions.includes(env.v)) return reject('unsupported_version');
    if (env.to !== this.#self) return reject('wrong_recipient');
    if (!isSignalingType(env.type)) return reject('unknown_type');

    let key: CryptoKey | null;
    let payload: unknown = undefined;
    if (env.type === 'PAIR_REQUEST') {
      if (!this.#acceptPair) return reject('unknown_sender', 'pairing not open');
      try {
        payload = decodePayload(env.payload);
      } catch {
        return reject('invalid_payload');
      }
      const pv = validateSignalingPayload('PAIR_REQUEST', payload);
      if (!pv.valid) return reject('invalid_payload', pv.errors[0]);
      let raw: Uint8Array;
      let derivedId: string;
      try {
        raw = b64urlDecode((payload as { public_key: string }).public_key);
        derivedId = await deviceIdFromPublicKey(raw);
      } catch {
        return reject('invalid_payload', 'public_key');
      }
      if (derivedId !== env.from) return reject('unknown_sender', 'public_key does not match from');
      try {
        key = await importPublicKey(raw);
      } catch {
        return reject('invalid_payload', 'public_key');
      }
    } else {
      key = await this.#resolveKey(env.from);
      if (!key) return reject('unknown_sender');
    }

    if (!(await verifyEnvelopeSignature(env, key))) return reject('bad_signature');

    if (Math.abs(this.#now() - env.ts) > this.#maxSkew) return reject('stale_timestamp');

    const k = seqKey(env.from, env.session);
    if (env.seq <= (this.#lastSeq.get(k) ?? 0)) return reject('replayed');

    if (payload === undefined) {
      try {
        payload = decodePayload(env.payload);
      } catch {
        return reject('invalid_payload');
      }
      const pv = validateSignalingPayload(env.type, payload);
      if (!pv.valid) return reject('invalid_payload', pv.errors[0]);
    }

    // Record seq only for fully accepted messages. Replaying a rejected message gains nothing.
    this.#lastSeq.set(k, env.seq);
    return { ok: true, envelope: env, payload, senderKey: key };
  }

  /** Call on session cleanup so a future session with the same id (never expected) starts fresh. */
  forgetSession(from: string, session: string): void {
    this.#lastSeq.delete(seqKey(from, session));
  }

  /** Call on revocation. */
  forgetSender(from: string): void {
    for (const k of [...this.#lastSeq.keys()]) if (k.startsWith(`${from}|`)) this.#lastSeq.delete(k);
  }
}

File: apps/android/app/src/main/kotlin/com/prc/controller/protocol/Receiver.kt
package com.prc.controller.protocol

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.longOrNull
import java.security.PublicKey

/**
 * Receiver rules for signaling envelopes (spec section 6), applied in this order. Every failure is
 * final for that message. This mirrors the TypeScript and Swift receivers and is checked against
 * the same vectors, so all three agree on what is acceptable.
 */
sealed class ReceiveResult {
    data class Accepted(val envelope: Envelope, val payloadJson: String) : ReceiveResult()
    data class Rejected(val reason: String, val detail: String? = null) : ReceiveResult()
}

class EnvelopeReceiver(
    private val selfDeviceId: String,
    private val resolveKey: (String) -> PublicKey?,
    private val now: () -> Long = { System.currentTimeMillis() },
    private val maxSkewMs: Long = Envelope.MAX_CLOCK_SKEW_MS,
    private val acceptPairRequests: Boolean = false,
) {
    private val lastSeq = HashMap<String, Long>()

    fun receive(raw: String): ReceiveResult {
        val bytes = Encoding.utf8(raw)
        if (bytes.size > Envelope.MAX_ENVELOPE_BYTES) return ReceiveResult.Rejected("too_large")

        val obj = try {
            Envelope.json.parseToJsonElement(raw) as? JsonObject ?: return ReceiveResult.Rejected("malformed", "not an object")
        } catch (e: Exception) {
            return ReceiveResult.Rejected("malformed", "not json")
        }
        val env = shape(obj) ?: return ReceiveResult.Rejected("malformed", "field types")

        if (env.v !in Envelope.SUPPORTED_VERSIONS) return ReceiveResult.Rejected("unsupported_version")
        if (env.to != selfDeviceId) return ReceiveResult.Rejected("wrong_recipient")
        if (env.type !in Envelope.TYPES) return ReceiveResult.Rejected("unknown_type")

        var payloadJson: String? = null
        val key: PublicKey
        if (env.type == "PAIR_REQUEST") {
            // A pairing request carries its own key, so it is the one message from a stranger. The
            // key must hash to the sender id, and the host still asks its owner to approve.
            if (!acceptPairRequests) return ReceiveResult.Rejected("unknown_sender", "pairing not open")
            val json = try { env.payloadJson() } catch (e: Exception) { return ReceiveResult.Rejected("invalid_payload") }
            val request = try {
                Envelope.json.decodeFromString(PairRequestPayload.serializer(), json)
            } catch (e: Exception) {
                return ReceiveResult.Rejected("invalid_payload", "PAIR_REQUEST")
            }
            val raw65 = try { Encoding.b64urlDecode(request.public_key) } catch (e: Exception) {
                return ReceiveResult.Rejected("invalid_payload", "public_key")
            }
            val derived = try { Identity.deviceId(raw65) } catch (e: Exception) {
                return ReceiveResult.Rejected("invalid_payload", "public_key")
            }
            if (derived != env.from) return ReceiveResult.Rejected("unknown_sender", "public_key does not match from")
            key = try { Identity.publicKey(raw65) } catch (e: Exception) {
                return ReceiveResult.Rejected("invalid_payload", "public_key")
            }
            payloadJson = json
        } else {
            key = resolveKey(env.from) ?: return ReceiveResult.Rejected("unknown_sender")
        }

        val sig = try { Encoding.b64urlDecode(env.sig) } catch (e: Exception) {
            return ReceiveResult.Rejected("bad_signature")
        }
        if (!Identity.verify(key, env.signingInput(), sig)) return ReceiveResult.Rejected("bad_signature")

        if (Math.abs(now() - env.ts) > maxSkewMs) return ReceiveResult.Rejected("stale_timestamp")

        val seqKey = "${env.from}|${env.session}"
        if (env.seq <= (lastSeq[seqKey] ?: 0L)) return ReceiveResult.Rejected("replayed")

        if (payloadJson == null) {
            val json = try { env.payloadJson() } catch (e: Exception) { return ReceiveResult.Rejected("invalid_payload") }
            if (!payloadValid(env.type, json)) return ReceiveResult.Rejected("invalid_payload", env.type)
            payloadJson = json
        }

        // Only a fully accepted message advances the counter; replaying a rejected one gains nothing.
        lastSeq[seqKey] = env.seq
        return ReceiveResult.Accepted(env, payloadJson)
    }

    fun forgetSession(from: String, session: String) {
        lastSeq.remove("$from|$session")
    }

    /**
     * The shape the envelope schema describes. A field of the wrong form is malformed, which is a
     * different answer from a signature that fails to verify: a 64-byte signature encodes to
     * exactly 86 base64url characters, so anything else never reaches the verifier.
     */
    private fun shape(obj: JsonObject): Envelope? {
        fun str(name: String): String? = (obj[name] as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull
        fun num(name: String): Long? = (obj[name] as? JsonPrimitive)?.takeIf { !it.isString }?.longOrNull
        val v = (obj["v"] as? JsonPrimitive)?.takeIf { !it.isString }?.intOrNull ?: return null
        val session = str("session") ?: return null
        return Envelope(
            v = v,
            type = str("type")?.takeIf { it.matches(TYPE_NAME) } ?: return null,
            from = str("from")?.takeIf { it.matches(DEVICE_ID) } ?: return null,
            to = str("to")?.takeIf { it.matches(DEVICE_ID) } ?: return null,
            session = session.takeIf { it.isEmpty() || it.matches(BYTES16) } ?: return null,
            seq = num("seq")?.takeIf { it >= 1 } ?: return null,
            ts = num("ts")?.takeIf { it >= 0 } ?: return null,
            payload = str("payload")?.takeIf { it.matches(BASE64URL) } ?: return null,
            sig = str("sig")?.takeIf { it.matches(SIGNATURE) } ?: return null,
        )
    }

    private fun payloadValid(type: String, json: String): Boolean {
        val j = Envelope.json
        return try {
            when (type) {
                "SESSION_REQUEST" -> j.decodeFromString(SessionRequestPayload.serializer(), json)
                "SESSION_CHALLENGE" -> j.decodeFromString(SessionChallengePayload.serializer(), json)
                "SESSION_AUTH" -> j.decodeFromString(SessionAuthPayload.serializer(), json)
                "SESSION_ACCEPT" -> j.decodeFromString(SessionAcceptPayload.serializer(), json)
                "SESSION_REJECT" -> j.decodeFromString(SessionRejectPayload.serializer(), json)
                "SESSION_END" -> j.decodeFromString(SessionEndPayload.serializer(), json)
                "PAIR_RESULT" -> j.decodeFromString(PairResultPayload.serializer(), json)
                // Media messages are carried through untouched until the video half is built.
                else -> j.parseToJsonElement(json) as? JsonObject ?: throw ProtocolException("not an object")
            }
            true
        } catch (e: Exception) {
            false
        }
    }

    private companion object {
        val DEVICE_ID = Regex("^[0-9a-f]{64}$")
        val TYPE_NAME = Regex("^[A-Z_]{1,32}$")
        val BYTES16 = Regex("^[A-Za-z0-9_-]{22}$")
        val SIGNATURE = Regex("^[A-Za-z0-9_-]{86}$")
        val BASE64URL = Regex("^[A-Za-z0-9_-]*$")
    }
}
